libmultilabel/nn/datasets_AttentionXML.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import annotations
+
+from typing import Sequence, Optional
+
+import numpy as np
+import torch
+from numpy import ndarray
+from scipy.sparse import csr_matrix, issparse
+from torch import Tensor, is_tensor
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+
+class PlainDataset(Dataset):
+    """Plain (compared to nn.data_utils.TextDataset) dataset class for multi-label dataset.
+    WHY EXISTS: The reason why this class is necessary is that it can process labels in sparse format, while TextDataset
+    does not.
+    Moreover, TextDataset implements multilabel binarization in a mandatory way. Nevertheless, AttentionXML already does
+    this while generating clusters. There is no need to do multilabel binarization again.
+
+    Args:
+        x: texts
+        y: labels
+    """
+
+    def __init__(self, x: list | ndarray | Tensor, y: Optional[csr_matrix | ndarray | Tensor] = None):
+        if y is not None:
+            assert len(x) == y.shape[0], "Sizes mismatch between texts and labels"
+        self.x = x
+        self.y = y
+
+    def __getitem__(self, idx: int) -> tuple[Sequence, ndarray] | tuple[Sequence]:
+        item = {"text": self.x[idx]}
+
+        # train/valid/test
+        if self.y is not None:
+            if issparse(self.y):
+                y = self.y[idx].toarray().squeeze(0)
+            elif is_tensor(self.y) or isinstance(self.y, (ndarray, torch.Tensor)):
+                y = self.y[idx]
+            else:
+                raise TypeError(
+                    "The type of y should be one of scipy.csr_matrix, torch.Tensor, and numpy.ndarry."
+                    f"Instead, got {type(self.y)}."
+                )
+            item["label"] = y
+        return item
+
+    def __len__(self):
+        return len(self.x)
+
+
+class PLTDataset(PlainDataset):
+    """Dataset for model_1 of AttentionXML.
+
+    Args:
+        x: texts
+        y: labels
+        num_classes: number of classes.
+        mapping: mapping from clusters to labels. Shape: (len(clusters), cluster_size).
+        clusters_selected: sampled predicted clusters from model_0. Shape: (len(x), predict_top_k).
+        cluster_scores: corresponding scores. Shape: (len(x), predict_top_k)
+    """
+
+    def __init__(
+        self,
+        x,
+        y: Optional[csr_matrix | ndarray] = None,
+        *,
+        num_classes: int,
+        mapping: ndarray,
+        clusters_selected: ndarray | Tensor,
+        cluster_scores: Optional[ndarray | Tensor] = None,
+    ):
+        super().__init__(x, y)
+        self.num_classes = num_classes
+        self.mapping = mapping
+        self.clusters_selected = clusters_selected
+        self.cluster_scores = cluster_scores
+        self.label_scores = None
+
+        # labels_selected are positive clusters at the current level. shape: (len(x), cluster_size * top_k)
```

Comment:
```suggestion
        # labels_selected are labels extracted from cluster_selected. shape: (len(x), cluster_size * top_k)
```

"top_k" isn't a var in this place. can roughly say
(size of each cluster) * (# selected clusters)